### ISSUE TITLE
[REQ] control trainer resistance from Garmin watch

### ISSUE THREAD

**Issue by stewpend0us** [cagnulein/qdomyos-zwift#1930]
**Is your feature request related to a problem? Please describe.**
I'd like my Garmin watch to be able to control the bike resistance based on a workout

**Describe the solution you'd like**
Seems like the easiest (without knowing anything about the code or constraints #953 ) thing to do would be to implement the FE-C ant+ device. Here's a repo that seems to have it working in c++ https://github.com/alex-hhh/TrainerControl/blob/master/src/FitnessEquipmentControl.cpp

**Describe alternatives you've considered**
Could probably get this working using the QZ Companion app with 'Background' processes. The QZ phone app would have to essentially poll the watch to get the current/next workout steps and then set the resistance that way. This seems more clunky to me and requires the QZ Companion app to be installed but maybe gets around some other technical issues I don't understand

**Comment by cagnulein** [comment]
duplicate of #1841 :)

**Comment by cagnulein** [comment]
yes i just need the time to do it :)

**Comment by stewpend0us** [comment]
Ah. I apologize for the duplicate! I might be able to help some if you can point me in the right direction...

**Comment by cagnulein** [comment]
no problem! do you need it on android right?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 30 dic 2023 alle 22:22 Stewart ***@***.***> ha
scritto:

**Comment by stewpend0us** [comment]
Yeah I use android. Just got an ant+ dongle so I can get cadence/power now!

**Comment by cagnulein** [comment]
great! are you a java developer? the code should be easy to do in case

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 30 dic 2023 alle 22:31 Stewart ***@***.***> ha
scritto:

**Comment by stewpend0us** [comment]
Not exactly but may be able to get it figured out. Is there a specific file/files I should look at to get a better idea?

**Comment by stewpend0us** [comment]
I do c++ development professionally and have done a couple "hello world" Android apps. I also don't give up easy...

**Comment by cagnulein** [comment]
@stewpend0us you are "another me" so :D i'm a firmware guy :) 

you can check this file to start 
https://github.com/cagnulein/qdomyos-zwift/blob/master/src/android/src/PowerChannelController.java

it's the power sensor on ant+ of qz

you can use it for reference.

you can create a PR so the CI on github will create directly a binary every change you made so you don't need to setup a build environment (only an editor :) )

Let me know if you have any questions

**Comment by stewpend0us** [comment]
Awesome. I'll take a closer look tonight and open a PR if it seems like something I can manage.

**Comment by cagnulein** [comment]
@stewpend0us any news?

**Comment by stewpend0us** [comment]
No news. Been traveling for work. I did read through the documentation for FE-C and it seems doable but need to find a minute

**Comment by cagnulein** [comment]
No worries, thanks for the update!


Il giorno gio 11 gen 2024 alle 11:33 Stewart ***@***.***> ha
scritto:

**Comment by stewpend0us** [comment]
Just curious...why aren't you using this: https://github.com/ant-wireless/ANT-Android-SDKs/tree/master/ANT%2B_Android_SDK

**Comment by cagnulein** [comment]
we are using that one :) it's the lib under qz

**Comment by stewpend0us** [comment]
I see the Ant lib but not the Ant+ lib?

**Comment by cagnulein** [comment]
yes i'm using only ant, if you think that it could be useful and easier
using the ant+ please go on! i never investigated it

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 13 gen 2024 alle 20:39 Stewart ***@***.***> ha
scritto:

**Comment by stewpend0us** [comment]
Seems like the ant+ lib might have been useful for some of the simpler stuff like heart rate...but doesn't add anything for faking a FE-C Trainer device. I'm working locally on my machine with just a place holder Android project just so I can check if things compile...once I have something worth trying out I'll put it here.

**Comment by cagnulein** [comment]
Cool, let me know.


Il giorno mar 16 gen 2024 alle 04:24 Stewart ***@***.***> ha
scritto:
